ocaml/xapi/xapi_vtpm.ml

```diff
@@ -12,19 +12,43 @@
    GNU Lesser General Public License for more details.
  *)
 
+let assert_no_vtpm_associated ~__context vm =
+  match Db.VM.get_VTPMs ~__context ~self:vm with
+  | [] ->
+      ()
+  | vtpms ->
+      let uuids =
+        List.map (fun v -> Db.VTPM.get_uuid ~__context ~self:v) vtpms
+      in
+      raise Api_errors.(Server_error (vtpm_max_amount_reached, uuids))
+
+let assert_vm_is_halted ~__context vm =
+  let attachable = [`Halted] in
+  match Db.VM.get_power_state ~__context ~self:vm with
+  | state when List.exists (( = ) state) attachable ->
+      ()
+  | state ->
+      let to_string = Record_util.power_to_string in
+      let expected = String.concat ", " (List.map to_string attachable) in
+      let error_params = [Ref.string_of vm; expected; to_string state] in
+      raise Api_errors.(Server_error (vm_bad_power_state, error_params))
+
 let introduce ~__context ~uuid ~vM ~profile ~contents =
   let ref = Ref.make () in
   Db.VTPM.create ~__context ~ref ~uuid ~vM ~profile ~contents ;
   ref
 
 let create ~__context ~vM =
+  assert_no_vtpm_associated ~__context vM ;
+  assert_vm_is_halted ~__context vM ;
   let uuid = Uuid.(to_string (make ())) in
   let profile = Db.VM.get_default_vtpm_profile ~__context ~self:vM in
   let contents = Xapi_secret.create ~__context ~value:"" ~other_config:[] in
   let ref = introduce ~__context ~uuid ~vM ~profile ~contents in
   ref
 
 let destroy ~__context ~self =
+  assert_vm_is_halted ~__context (Db.VTPM.get_VM ~__context ~self) ;
   let secret = Db.VTPM.get_contents ~__context ~self in
   Db.Secret.destroy ~__context ~self:secret ;
   Db.VTPM.destroy ~__context ~self
```

Comment:
I assume xapi_db_gc will then go and delete the link from the VM to this vTPM?

Reply:
That's correct

Reply:
That's not what the GC does. If the link is defined a one-to-many relation, then if the "one" side is deleted, the other side is automatically updated by the database. A GC rule is needed for the case where the VM object is deleted, and its VTPM objects need to be automatically deleted. Do we have a rule for that?

Reply:
> Do we have a rule for that?

No, such rule would delete the tpm state with the current model

Reply:
I would think in currently that's true, I've created an issue to track that